File: Python/Easy/valid_parentheses.py
"""
Problem: Valid Parentheses
Link: https://leetcode.com/problems/valid-parentheses/

Check if the input string has valid parentheses pairing.
"""

def is_valid(s: str) -> bool:
    stack = []
    pairs = {')': '(', '}': '{', ']': '['}

    for char in s:
        # If it's a closing bracket
        if char in pairs:
            if not stack or stack[-1] != pairs[char]:
                return False
            stack.pop()
        else:
            # Opening bracket
            stack.append(char)

    return not stack  # True if empty


# ------------------------------------------------------------
# Quick test
# ------------------------------------------------------------
if __name__ == "__main__":
    tests = ["()", "()[]{}", "(]", "([)]", "{[]}"]
    for t in tests:
        print(f"{t:8} -> {is_valid(t)}")

File: CSharp/Easy/ValidParentheses.cs
using System;
using System.Collections.Generic;

public class Solution
{
    public static bool IsValid(string s)
    {
        // Stack to keep track of opening brackets
        Stack<char> stack = new();

        // Mapping of closing brackets to their corresponding opening ones
        Dictionary<char, char> pairs = new()
        {
            { ')', '(' },
            { '}', '{' },
            { ']', '[' }
        };

        // Iterate through each character in the string
        foreach (char c in s)
        {
            // If it's a closing bracket
            if (pairs.ContainsKey(c))
            {
                // Check if stack is empty or the top element doesn't match
                if (stack.Count == 0 || stack.Peek() != pairs[c])
                    return false;

                // Valid pair found -> pop from stack
                stack.Pop();
            }
            else
            {
                // It's an opening bracket -> push onto stack
                stack.Push(c);
            }
        }

        // If the stack is empty, all brackets are properly closed
        return stack.Count == 0;
    }

    // ------------------------------------------------------------
    // Quick test
    // ------------------------------------------------------------
    public static void Main()
    {
        string[] tests = { "()", "()[]{}", "(]", "([)]", "{[]}" };

        foreach (var t in tests)
        {
            Console.WriteLine($"{t,-8} -> {IsValid(t)}");
        }
    }
}
